File: python/crosswordSolver.py
from crosswordMaker import Crossword

def crosswordSolver (crossword):
    index = 0
    while True:
        if index < 0:
            raise ValueError('No solution')
            
        backtrack(crossword, index)

        crossword.solutions[index] = crossword.possibleMatches[index][0]
        crossword.possibleMatches[index] = crossword.possibleMatches[index][1:]
        
        if checkSolutions(crossword):
            if all(crossword.solutions):
                solution = ''.join(crossword.solutions)
                return solution
            index += 1
        else:
            backtrack(crossword, index)

                
def checkSolutions(crossword):
    for i in xrange(crossword.height):
        start = crossword.width * i
        rowSolution = ''.join(crossword.solutions[start: crossword.width + start])
        print rowSolution
        key = 'Y' + str(i)
        
        clues = crossword.clues[key]
        
        for clue in clues:
            if not partialMatch(rowSolution, clue, crossword.width):
                return False
            
    for j in xrange(crossword.width):        
        colSolution = ''.join(crossword.solutions[j::crossword.width])
        print colSolution

        key =  'X' + str(j)
        
        clues = crossword.clues[key]
        
        for clue in clues:
            print (partialMatch(colSolution, clue, crossword.height))
            if not partialMatch(colSolution, clue, crossword.height):
                return False; 

    return True


def partialMatch(string, exp, maximum):
    print string
    print exp
    partial = len(string) != maximum
    regexObj = re.compile(exp)

    if regexObj.fullmatch(string, partial=partial):
        return True
    return False

def backtrack(crossword, index):
    if not crossword.possibleMatches[index]:
        crossword.solutions[index] = ''
        crossword.possibleMatches[index] = crossword.characters
        index -= 1

    

File: js/crosswordSolver.js
var Crossword = require('./crosswordMaker');

require('./partialMatchParser');


let characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ 1234567890!"#%&,<=>@~_`;:'


function crosswordSolver(crossword) {

    let index = 0;

    let cont = true;

    while (cont) {
        if (index < 0) {
            throw Error('I think there might have been a mistake.')
        }

        backtrack(crossword, index);

        crossword.solutions[index] = crossword.possibleMatches[index].slice(0, 1);
        crossword.possibleMatches[index] = crossword.possibleMatches[index].slice(1);

        if (checkSolutions(crossword)) {
            if (crossword.solutions.every(e => e.length !== 0)) {
                let solution = crossword.solutions.join('')
                return solution;
            }
            index++;
        }

        else { backtrack(crossword, index); }
    }

}


function checkSolutions(crossword) {
    //check all rows
    for (let i = 0; i < crossword.height; i++) {
        const start = crossword.width * i;
        const rowSolutionSlice = crossword.solutions.slice(start, crossword.width + start);
        const rowSolution = rowSolutionSlice.join('');

        let key = 'Y' + i;

        let clue = crossword.clues[key];
        //can't stop execution of function within a forEach, so we use a for loop

        for (let ii = 0; ii < clue.length; ii++) {
            if (!partialMatch(rowSolution, clue[ii])) {
                return false;
            }
        }
    }

    //check all columns
    for (let j = 0; j < crossword.width; j++) {
        const solutionsRay = crossword.solutions.slice(j).filter((element, index) => {
            return index % this.width === 0;
        })

        const colSolution = solutionsRay.join('');

        let key = 'X' + j;

        let clue = crossword.clues[key];
        for (let jj = 0; jj < clue.length; jj++) {
            if (!partialMatch(colSolution, clue[jj])) {
                return false;
            }
        }
    }

    return true;
}

function partialMatch(str, exp) {

    var regexObj = new RegExp(exp);
    var partialMatchRegex = regexObj.toPartialMatchRegex();
    var result = partialMatchRegex.exec(str);
    var matchType = regexObj.exec(str) ? 'Full match' : result && result[0] ? 'Partial match' : 'No match';

    if (str === '' || (matchType === 'Full match' || matchType === 'Partial match')) {
        return true;
    }

    return false;
}

function backtrack(crossword, index) {
    if (!crossword.possibleMatches[index]) {
        crossword.solutions[index] = '';
        crossword.possibleMatches[index] = characters;
        index--;
    }
}
